### DISCUSSION TITLE
Refresh Page after URL parameters changes

### DISCUSSION THREAD

**Discussion by Eifhen** [solidjs/solid#1745] [Q&A]
Currently im working on a page and i have the following problem:
i have a route as the following  "/products/1/detail"
i have an objects array that simulates a database and when i render the page, i go throw the objects array and i get the object that has the same id that it is in the route parameter. 

That works fine. but the thing is that in the same page i have a section where i show lets say other products that the user can see, and when the user clicks on one of those products he will be redirect to the detail of that product, by doing that the parameter in the url will change. for example:
from "/products/1/detail" to "/products/2/detail", the thing is that when i do that the url will change but the data in the page will not be updated.
is there a way to re-render de page when the url parameters change?
i asume that i could have the same problem if instead of searching inside an object i fetch data from an API depending on the route parameter.

I apologize for my bad English, but I am not a native speaker.

**Comment by palatam** [comment]
I have the exact same problem.
All the solutions I have found so far simply don't work.
It works with a single parameter and createResource but no solution for multiple parmeters (say /products/:id/:aspect) ... the URL changes but page stays the same and nothing gets triggered (tried createResource, tried createEffect, tried derived signals, ...)

**Comment by ryansolid** [comment]
I'd look to see how you are accessing the params. The params object is a proxy of sorts so.. doing `params.___` need to accessed under the reactive scope. This should all work so would request an example of it not working.

**Reply by Sheldor5** [reply]
just like https://docs.solidjs.com/guides/routing-and-navigation#accessing-parameters 2nd example where it says "Every time the id parameter changes in this example, the fetchUser function is called to fetch the new user data.".

This works with a single parameter but how do I do the same if the route has 2 parameters like "/licences/:context/:id" ?? here I have parameters "context" (e.g. "product" or "customer") and "id" and in the HTML I have a <A href=" /licences/" + otherContext + "/" + item.id + "">... the click on the link changes the URL which is the same route as per page/route as per "/licences/:context/:id" -> but the page does not re-render or triggers any reactive objects ...

So do you have a working example or any idea on how to "re-render" the page if all (more than one) parameters of the current route changes?

example:

```
current route: /licences/customer/1

with <A href="/licences/product/1">...</A> (id is the same just coincidentally)
```

click on the link changes the URL but does not trigger any reactive code nor re-rendering logic

**Comment by nlfiedler** [comment]
Sorry I'm a little late to the party, but this is a short example that does not work as expected:

```
function AssetDetails() {
  const params = useParams();
  const client = useApolloClient();
  const [assetQuery] = createResource(params, async (params) => {
    const { data } = await client.query({
      query: GET_ASSET,
      variables: { id: params.id! }
    });
    return data;
  });

  return (
    <Suspense fallback={<button class="button is-loading">...</button>}>
      <div>
        <Show when={assetQuery()} fallback="Loading...">
          <AssetFigure asset={assetQuery()?.asset!} />
          <AssetForm asset={assetQuery()?.asset!} />
        </Show>
      </div>
    </Suspense>
  );
}
```

When navigating to `/asset/:id` the first time, the page loads the resource identified by `params.id`. However, later the user presses a button and a `redirect('/asset/newid')` is called and... nothing happens. If I add a `createEffect()` to log the `params.id` value, I see the change, but the page itself does not refresh.

Any guidance would be appreciated, thank you.

**Comment by nlfiedler** [comment]
I am using this work-around, in case it helps anyone else:

```javascript
  const params = useParams();
  const client = useApolloClient();
  const [assetQuery, { refetch }] = createResource(
    () => params.id,
    async (assetId) => {
      const { data } = await client.query({
        query: GET_ASSET,
        variables: { id: assetId }
      });
      return data;
    }
  );
  const location = useLocation();
  createEffect(() => refetch(location.pathname));
```
